### ISSUE TITLE
npm token create password (npm token)

### ISSUE THREAD

**Issue by tvb** [verdaccio/verdaccio#541]
```
$ npm token create --registry=http://my_registry:5000
npm password:
```

What is the password asking its for?

**Comment by juanpicado** [comment]
@tvb  ???

**Comment by tvb** [comment]
I am trying to create a token on the server itself as I still have issues getting authenticated by login in.

**Comment by juanpicado** [comment]
`verdaccio` does not support `npm token` endpoint yet.

**Comment by juanpicado** [comment]
But it sounds like a good research for somebody is willing to contribute 👍

**Comment by juanpicado** [comment]
Just login with verdaccio and get it from `.npmrc` file in your home directory.

**Comment by tvb** [comment]
I can't get it to work. Whenever I login this ends up in my .npmrc file:
```
@scope:registry=http://mydomain.com:5000/
//mydomain.com:5000/:_password=xxxxx
//mydomain.com:5000/:username=myusername
//mydomain.com:5000/:email=myemail@mydomain.com
//mydomain.com:5000/:always-auth=false
```

**Comment by juanpicado** [comment]
@tvb https://github.com/npm/npm/issues/19652 ?

**Comment by juanpicado** [comment]
Options: 
1. Downgrade `npm`.
2. Use `yarn`.

Let us know 👍

**Comment by mikew** [comment]
Using an older version of node via Docker is also an option, just not a great one.

```
$ docker run --rm -it node:9.2.0-alpine sh
/ # npm login --registry https://npm.company.com
Username: admin
Password:
Email: (this IS public) admin@company.com
Logged in as admin on https://npm.company.com/.
/ # cat ~/.npmrc
//npm.company.com/:_authToken=Rwl9t+GHjlgP+brFJ6WycIe1y6r3Z+ShUEqsLusmFC11w3n6ex8JdmkMoKv/0U/D
/ #
```

**Comment by varijkapil13** [comment]
Hi, this issue is in Pick up for v4, I would love to pick it up and contribute...

**Comment by juanpicado** [comment]
@varijkapil13 sure, all yours. I can drive you along the process if you need help. Let me know what would be your action plan to make it real.

**Comment by varijkapil13** [comment]
@juanpicado Thanks.  I think I will need some help getting started with the code. Regarding action plan, I will have to some research how to implement it, and understand what npm does on.
I just joined the verdaccio discord server. It will probably be better if we discuss this there.

**Comment by juanpicado** [comment]
Hi @varijkapil13 , any update? We are about to finish https://github.com/verdaccio/verdaccio/pull/1034 and the next step for me is `npm create token`. Can I help you with something?

**Comment by varijkapil13** [comment]
Hi @juanpicado no, no updayes yet. I hadn't had that much time to invest into this. Plus I need some time getting used to express and nodejs. Feel free to take it over. I make sure to follow the progress and learn on the way.
In the meantime maybe I can help with smaller stuff, like refactoring and such.

**Comment by juanpicado** [comment]
@varijkapil13 sure, don't hesitate to contribute :) I'm here to help and provide guidance for anyone.

**Comment by varijkapil13** [comment]
@juanpicado thanks! i think I'll stick to the smaller parts for now. Implementing authentication stuff would take way more time for me. Since I fail to understand the workings of npm requests and how they are handled.

I'll take care of some things from #973 .
(Will open a PR soon for 

`Get rid off require('../lib/logger'); in favor of ES modules import xxx from xxx.`
`Replace import * as Utils from '../../lib/utils'; by destructuring format import {xxx, yyy} from '../../lib/utils';)`
)

**Comment by juanpicado** [comment]
I'm trying to give a bit of shape to this feature.
There are a couple of topics I'd like to describe and looking for opinions.

### package manager support
* npm ✅ 
* pnpm ✅ 
* yarn ❌ 

### package manager support
* npm ✅ 
* pnpm ✅ 
* yarn ❌ 

### commands

```
➜ npm token --help       
npm token list
npm token revoke <tokenKey>
npm token create [--read-only] [--cidr=list]
```

#### npm token list

First a bit of context how is, in the next major release, we support 2 different ways to generate tokens, one is *legacy* which is the currently used on `latest` and `JWT`. (for deep context https://github.com/verdaccio/verdaccio/pull/896) The idea is support both systems until `legacy` is being removed in future major releases.

Verdaccio does not store tokens anywhere, thus, once token is signed and delivered to the client, we lose the control of it. Without token storage, there is no way to support `npm token list`. I'm looking a safe way to address this process.

`npm token revoke` and `npm token create` are not hard to handle once the first subject is done.

**Comment by JanneHarju** [comment]
We need this for CI/CD builds. Please add this soon.

**Comment by juanpicado** [comment]
PR are open 👍

**Comment by JanneHarju** [comment]
![image](https://user-images.githubusercontent.com/17338490/54282051-c9415a80-45a3-11e9-87fd-f421a79da78d.png)
Just wondering what these removes mean. Will this be in Verdaccio 4?

**Comment by juanpicado** [comment]
@JanneHarju no it's too big and at least I won't work on it for now.

**Comment by Eomm** [comment]
Hi,
I have checked to implement this feature and I think it should be necessary:

- implement some endpoint in `src/api/endpoint/api` that reply to:
  - `GET https://example.it/-/npm/v1/tokens` to get the token list
  - `POST https://example.it/-/npm/v1/tokens` to create the a token
  - need to find what is necessary for the `revoke` command
- use the `IStorageHandler` to store these tokens (+ read-only flag + IPs cidr)
- some test

@juanpicado 
Do you think that it is enough or I'm missing something?
Do you know if there are some docs that explain better the endpoint to expose?
Do you think that it would be possible to release this feature in next semver-minor v3 when it will be implemented?

Thank you

**Comment by juanpicado** [comment]
hi @Eomm 

> implement some endpoint in src/api/endpoint/api that reply to:

yes, following the same structure would be located there eg: `token.js` 

> GET https://example.it/-/npm/v1/tokens to get the token list
> POST https://example.it/-/npm/v1/tokens to create the a token
> need to find what is necessary for the revoke command

I'm not 100% sure, but I'll check that as well. 

> use the IStorageHandler to store these tokens (+ read-only flag + IPs cidr)

Yes, here is the tricky part. We clearly need to modify the `IStorageManager` and the `local-storage`, so, the plugin IMHO should hold the responsibility for handle persistence of tokens. This will impact current plugins so we need to be flexible there, plugins depend on external developers and it will take time until they align with our new API. About security wise, well, I'd love to store tokens in a safe way, let me know if you have ideas in this area.

> Do you think that it is enough or I'm missing something?

Not 100% sure, but, we will figure out :-)

> Do you know if there are some docs that explain better the endpoint to expose?

Unfortunately no, npmjs is a closed API, so, reverse engineering is required here. My approach is just check the `npm` client code, all is there.

> Do you think that it would be possible to release this feature in next semver-minor v3 when it will be implemented?

I'm not accepting features on `v3` anymore, only bug fixing. My reasons are, 1. the stability of the product, 2. people using storage plugins will want to use it and as I said above, I cannot force storage plugins to align to this, but, those plugins are under our control, yes we can. `v4` is not that far away to be launched, so, I think to make sense to included it on `v4` instead. 

You can find me in our chat if you need more ideas or guidance in a quick way.

**Comment by Eomm** [comment]
> the plugin IMHO should hold the responsibility for handle persistence of tokens.

Sorry, I don't understand this part: do you suggest to build some plugins to archive this feature?
Because the hard part I see is sync all these steps:

+ one auth plugin: to read and validate tokens on actions
+ one storage plugin: to add the new functions to store the tokens
+ one middlewere: to expose the endpoints

> About security wise, well, I'd love to store tokens in a safe way, let me know if you have ideas in this area.

The simplest pattern I would approach in first place would be to add a setting for the master-password and then store the token as encrypted strings. If I will use standard Node.js API, it would require Node.js >= 10.5.0. [Here](https://github.com/Eomm/keepy-cli/blob/master/lib/CryptoString.js) I did some experiment in this way. Could be good for you this requirement?

> v4 is not that far away to be launched

Clear, I was thinking to build this feature in a fastest way because my company need it.
But I think I could archive the quick-and-dirty part with a auth plugin and some `const` 😅
Meanwhile building better version of this "must-have" feature 👍

**Comment by Eomm** [comment]
Hi @juanpicado 
I build [verdaccio-static-token](https://github.com/Eomm/verdaccio-static-token) plugin for now that fit my needs and maybe could be useful to others that are watching this issue.

Implementing that plugin I notice how I should implement the complete feature, I'll hope to start soon to write some code 👍

**Comment by juanpicado** [comment]
@Eomm nice job ! I'll check in detail your implementation, I just quickly overseen a bit looks interesting.

**Comment by carlosen14** [comment]
You can do it right now following these steps.. ( i think this might be the old way, but it is working for me on NPMJS)

- Your project will have this `.npmrc` file, you can also store it on git repo (i am doing it, i am not sure if it is wrong, anyway, it only has the repo url, it is locally and the repos are user-specific locked) 
```
#.npmrc
registry = http://127.0.0.1:4873`
```
- Generate a encoded token by username:password
```bash
#auth can be generate by username:password 
echo -n 'usernm:passwd' | openssl base64
#output YWRtaW46YWRtaW4xMjN=
```
- Store this string on a environment variable like depending your CI/CD.
- On CI/CD when you compile.
```bash
docker build -t verdaccio --build-arg NPMTOKEN=$NPMTOKEN .
```
```Dockerfile
RUN [[ -z "$NPMTOKEN" ]] && exit 1 || echo "_auth=${NPMTOKEN}" >> .npmrc
RUN npm ci
```
- This will be able to download the private ones..

:)

**Comment by juanpicado** [comment]
If you want to try a PoC, feel free to share feedback.

```
npm i -g verdaccio@feat-token --registry https://registry.verdaccio.org
```

**Comment by carlosen14** [comment]
> If you want to try a PoC, feel free to share feedback.
> 
> ```
> npm i -g verdaccio@feat-token --registry https://registry.verdaccio.org
> ```

How to use it?

**Comment by juanpicado** [comment]
@carlosen14 https://docs.npmjs.com/cli/token.html

**Comment by carlosen14** [comment]
> If you want to try a PoC, feel free to share feedback.
> 
> ```
> npm i -g verdaccio@feat-token --registry https://registry.verdaccio.org
> ```

You want me to install this on verdaccio directly? i'm using verdaccion on docker

**Comment by juanpicado** [comment]
> You want me to install this on verdaccio directly? i'm using verdaccio on docker.
I installed on local machine but continues asking for password and returns 404.

Docker image for testing is not available yet, but it will. I'll let you know @carlosen14 in the next week.

**Comment by carlosen14** [comment]
ok, i did `npm login --registry verdaccio.localhost:1234` and this saved a token on `/root/.npmrc` 

So, there is a token, i copied from there!
example: 
(REAL TOKEN IS WAY MUCH LONGER THAN EXAMPLE)
```
//verdaccio.localhost:1234/:_authToken=eyJh.TY0N.xZik8GL
```

On CICD i put it using
```
echo '//verdaccio.localhost:1234/:_authToken=eyJh.TY0N.xZik8GL"' >> `npm config get userconfig`
```

**Comment by carlosen14** [comment]
> > You want me to install this on verdaccio directly? i'm using verdaccio on docker.
> > I installed on local machine but continues asking for password and returns 404.
> 
> Docker image for testing is not available yet, but it will. I'll let you know @carlosen14 in the next week.

@juanpicado  is this released on 4.3?
